### ISSUE TITLE
duplicate entries in airlines_processed.csv

### ISSUE THREAD

**Issue by chijames** [asappresearch/dialog-intent-induction#13]
Hi,

I am trying to run the data construction code and I notice something weird. Specifically, in the file airlines_processed.csv, there are about 700 entries with duplicate tweet_id. To name a few, 638092, 827255, 394166...etc. Is this expected?

Thanks.

**Comment by hpasapp** [comment]
can you provide the exact sequence of instructions you are running please?

**Comment by chijames** [comment]
The above observation is made based on the current airlines_processed.csv in this repo.

**Comment by hpasapp** [comment]
Hmmmm, right:

```
3.7.7:3.8.3 ~/git/dialog-intent-induction/data (master|…6)$ grep '^638092,' airlines_processed.csv
638092,UNK,airlines.csv:2347,<cust__ any airlines having __twitter_tag__ deals ? ? __company__ __twitter_at__ __twitter_at__ __twitter_at__ __twitter_at__ __twitter_at___airways __cust>,<rep__ __cust__ contact our reservations team at __num__ - __num__ for the lowest fares we have available . __rep>
638092,UNK,airlines.csv:14378,<cust__ any airlines having __twitter_tag__ deals ? ? __twitter_at__ __twitter_at__ __company__ __twitter_at__ __twitter_at__ __twitter_at___airways __cust>,<rep__ __cust__ contact our reservations team at __num__ - __num__ for the lowest fares we have available . __rep>
638092,UNK,airlines.csv:26968,<cust__ any airlines having __twitter_tag__ deals ? ? __twitter_at__ __twitter_at__ __twitter_at__ __twitter_at__ __twitter_at__ __company__ __cust>,<rep__ __cust__ contact our reservations team at __num__ - __num__ for the lowest fares we have available . __rep>
```

Grepping the original twcs.csv, this line occurs only once:
```
3.7.7:3.8.3 ~/git/dialog-intent-induction/data (master|…6)$ grep '^638092,' ~/data/twcs.csv 
638092,271298,True,Wed Nov 22 19:05:30 +0000 2017,Any airlines having #blackfriday deals?? @Delta @116450 @AmericanAir @120910 @VirginAtlantic @British_Airways,638091,
```

However, I notice it contains multiple airlines. Is it possible that my processing code did something like:
```
rows = []
for airline in airlines:
    append all lines having that airline to rows
```
... meaning that any line containing multiple airlines occurs multiple times? ie, do all the duplicates contain multiple airlines?

Checking the next reference in your list, this appears consistent with this hypothesis, having both Delta and AmericanAir:
```
3.7.7:3.8.3 ~/git/dialog-intent-induction/data (master|…6)$ grep '^827255,' ~/data/twcs.csv 
827255,317056,True,Fri Oct 13 01:07:06 +0000 2017,Hey @Delta can you help @AmericanAir get us a status on AA2990? They seem to be struggling to locate the plane.,827252,
```

**Comment by chijames** [comment]
I guess it's related to https://github.com/asappresearch/dialog-intent-induction/blob/master/datasets/twitter_dataset_preprocess.py#L55 ?

**Comment by hpasapp** [comment]
yes, I think so... probably it should get eg the row ids or tweet ids into a set or dict, and then write those out, rather than just appending to a list.

**Comment by hpasapp** [comment]
(or else check each line against all companies)

**Comment by chijames** [comment]
I write a small script to postprocess the dataset:
```
res = {}
out = open('new.csv', 'w')
with open('airlines_processed.csv') as infile:
    for row in infile:
        tweet_id = row.split(',')[0]
        if tweet_id not in res:
            res[tweet_id] = row
        else:
            if row.split(',')[1] != 'UNK':
                res[tweet_id] = row
for k, v in res.items():
    out.write(v)
```
This ensures the dataset contains only unique tweet_id and all the 492 labeled instances are kept.

Then I use the new data to run the experiment again:

```
# qt pre-training
cmdlines/airlines_qt.sh
cmdlines/airlines_qt_mvc.sh
```

Here is the log:
```
loading dataset
num of class = 14
loading glove
loading oov
instantiating model
created randomly initialized model
vocab_size 33244
2021-01-08 01:52:16.935317 epoch 1, train_loss=4638.6070 test_acc=0.4094
2021-01-08 01:53:01.059375 epoch 2, train_loss=3123.3651 test_acc=0.4688
2021-01-08 01:53:45.193162 epoch 3, train_loss=2271.2106 test_acc=0.4803
2021-01-08 01:54:29.305628 epoch 4, train_loss=1655.5052 test_acc=0.4624
2021-01-08 01:55:13.622544 epoch 5, train_loss=1198.9955 test_acc=0.4566
loaded best state from epoch 3
applied post-pretraining
2021-01-08 01:55:41.576890 pretrain: test prec=0.5153 rec=0.4196 f1=0.4625 acc=0.4033
saved model to  data/airlines_qt.pth
```

After training for 50 epochs:
```
restoring model for best dev epoch 12
2021-01-08 02:10:57.173524 test prec=0.4807 rec=0.3931 f1=0.4325 acc=0.3788
```

This worries me a bit as the av-kmeans seems to lower the performance.

**Comment by yi-asapp** [comment]
the airline dataset is really small, you will notice that the results can be quite different if you change the seed etc.

remove some entries makes it even smaller, so the performance variance will increase as well.

I would 
a) get more tweets annotated, or
b) trust the ubuntu results more

**Comment by chijames** [comment]
Here are my thoughts and additional observations:

1. I try 3 additional seeds:
```
2021-01-08 04:19:35.447561 pretrain: test prec=0.4847 rec=0.3931 f1=0.4341 acc=0.3686
restoring model for best dev epoch 17
2021-01-08 04:34:46.827517 test prec=0.4786 rec=0.3992 f1=0.4353 acc=0.3768

2021-01-08 04:40:02.650110 pretrain: test prec=0.5214 rec=0.3951 f1=0.4495 acc=0.3625
restoring model for best dev epoch 34
2021-01-08 04:55:22.606520 test prec=0.5173 rec=0.3747 f1=0.4346 acc=0.3564

2021-01-08 04:59:55.087829 pretrain: test prec=0.4868 rec=0.4033 f1=0.4411 acc=0.3809
restoring model for best dev epoch 15
2021-01-08 05:15:13.911388 test prec=0.5092 rec=0.4033 f1=0.4501 acc=0.3971
```

Seems like qt pretraining itself is good enough for clustering. I don't see the need of proposed av-kmeans here.

2. I only remove 700 instances of the data, which is about 1.6%. This shouldn't be a problem. Also, the sizes of these two datasets are 104479 (ask_ubuntu) vs 42313 (twitter_airline). If twitter dataset is considered "really small", a dataset (ask_ubuntu) which is only 2.5 larger shouldn't be considered "large". But yes, I will check the ubuntu dataset as it is indeed larger.

3. Annotating more tweets would definitely help. Looking forward to a larger dataset :)
